### ISSUE TITLE
Poor support for generating large-ish data sets (with example provided)

### ISSUE THREAD

**Issue by epgui** [HypothesisWorks/hypothesis#3493]
I've been trying to adopt hypothesis to generate plausible data sets (which are transformed into api responses and then used for integration testing).

I am used to doing generative testing with clojure.spec.alpha, clojure.spec.gen.alpha and clojure.test.check, and I have never had any issues whatsoever with data size / performance with these libraries.

For "large" (not so large in my opinion) data sets, using hypothesis can be very frustrating, and the output is not very helpful.

I've reproduced below a typical, minimal example of what I'm trying to do:

```python
from datetime import datetime
from hypothesis import given, HealthCheck, settings, Verbosity, strategies as st
from hypothesis.extra.pandas import column, data_frames, range_indexes

# Constraining range solely for performance reasons
st_integers = st.integers(min_value=0, max_value=1000000)


@st.composite
def gen_dt_year_month(draw):
    """A strategy for datetimes truncated to the beginning of the month."""
    return draw(
        st.datetimes(
            # Constraining range solely for performance reasons
            min_value=datetime.fromisoformat("2010-01-01"),
            max_value=datetime.today(),
        )
    ).replace(day=1, hour=0, minute=0, second=0, microsecond=0)


@st.composite
def my_data_set_strategy(draw, num_rows):
    cols = [
        # Not even bothering with strings yet, I figure integers would be faster
        column('company_name', elements=st_integers, fill=st.nothing(), dtype=int),
        column('date', elements=gen_dt_year_month(), fill=st.nothing()),
        column('field_a', elements=st_integers, fill=st.nothing(), dtype=int),
        column('field_b', elements=st_integers, fill=st.nothing(), dtype=int),
        column('field_c', elements=st_integers, fill=st.nothing(), dtype=int),
        column('field_d', elements=st_integers, fill=st.nothing(), dtype=int),
        column('field_e', elements=st_integers, fill=st.nothing(), dtype=int),
        column('field_f', elements=st_integers, fill=st.nothing(), dtype=int),
        column('field_g', elements=st_integers, fill=st.nothing(), dtype=int),
        column('field_h', elements=st_integers, fill=st.nothing(), dtype=int),
        column('field_i', elements=st_integers, fill=st.nothing(), dtype=int),
        column('field_j', elements=st_integers, fill=st.nothing(), dtype=int),
        column('field_k', elements=st_integers, fill=st.nothing(), dtype=int),
        column('field_l', elements=st_integers, fill=st.nothing(), dtype=int),
        column('field_m', elements=st_integers, fill=st.nothing(), dtype=int),
        column('field_n', elements=st_integers, fill=st.nothing(), dtype=int),
        column('field_o', elements=st_integers, fill=st.nothing(), dtype=int),
        column('field_p', elements=st_integers, fill=st.nothing(), dtype=int),
        column('field_q', elements=st_integers, fill=st.nothing(), dtype=int),
        column('field_r', elements=st_integers, fill=st.nothing(), dtype=int),
        column('field_s', elements=st_integers, fill=st.nothing(), dtype=int),
        column('field_t', elements=st_integers, fill=st.nothing(), dtype=int),
        column('field_u', elements=st_integers, fill=st.nothing(), dtype=int),
        column('field_v', elements=st_integers, fill=st.nothing(), dtype=int),
        column('field_w', elements=st_integers, fill=st.nothing(), dtype=int),
        column('field_x', elements=st_integers, fill=st.nothing(), dtype=int),
        column('field_y', elements=st_integers, fill=st.nothing(), dtype=int),
        column('field_z', elements=st_integers, fill=st.nothing(), dtype=int),
    ]
    return draw(
        data_frames(
            columns=cols,
            index=range_indexes(min_size=num_rows, max_size=num_rows),
        ).filter(
            # Eliminate dataframes that have all-zero values because those are not
            # important to consider in the context of our tests.
            lambda df: set(df["field_a"]) != {0}
        )
    )


@given(my_data_set_strategy(num_rows=50))
@settings(
    # Turning this down aggressively for performance reasons
    max_examples=5,
    # These health checks are so easily triggered...
    suppress_health_check=[
        HealthCheck.large_base_example,
        HealthCheck.data_too_large,
        HealthCheck.too_slow,
    ],
    # Please tell me more!
    verbosity=Verbosity.verbose,
)
def thing(data):
    print(data)
    assert True


thing()
```

For `num_rows=50`, it takes 704ms to generate a data set, which is slow but not _completely_ unreasonable.

However, as soon as I set `num_rows=100`, the wheels start spinning and spinning (>1 minute) and I get the following output:

```
You can add @seed(275777655191065385259004160039161633348) to this test to reproduce this failure.
---------------------------------------------------------------------------
Unsatisfiable                             Traceback (most recent call last)
Cell In [170], line 11
      8     print(data)
      9     assert True
---> 11 thing()

Cell In [170], line 2, in thing()
      1 @with_timing()
----> 2 @given(my_data_set_strategy())
      3 @settings(
      4     max_examples=5,
      5     suppress_health_check=[HealthCheck.large_base_example, HealthCheck.data_too_large, HealthCheck.too_slow],
      6     verbosity=Verbosity.verbose,
      7 )
      8 def thing(data):
      9     print(data)
     10     assert True

    [... skipping hidden 1 frame]

File /usr/local/lib/python3.9/site-packages/hypothesis/core.py:840, in StateForActualGivenExecution.run_engine(self)
    838     if runner.valid_examples == 0:
    839         rep = get_pretty_function_description(self.test)
--> 840         raise Unsatisfiable(f"Unable to satisfy assumptions of {rep}")
    842 if not self.falsifying_examples:
    843     return

Unsatisfiable: Unable to satisfy assumptions of thing
```

It's unclear to me what assumption it's unable to satisfy: the only constraint I have is for all values in the dataframe to not be zero-valued, which I would expect would be no problem after the first handful (<5) of examples.

Generating "large" (not so large in my opinion) dataframes is not the only situation in which I've run into this frustrating type of result. Previously I was generating a list of dicts with complex constraints and relationships between items... It was my hope that generating a dataframe with simpler constraints would reduce the chance that I was doing something silly.

I really want to love this library, and I want to be an advocate for its adoption at work, but it's so difficult to understand what's happening or why some strategies don't work very well, that we're very close to abandoning its use. I'm hoping someone can help point me in the right direction!

---

Reproducibility:

- Mac OS 12.6 (Apple M1 Pro), with everything running in a docker container locally
- python: 3.9.9
- hypothesis: 6.54.3

**Comment by langfield** [comment]
> I really want to love this library, and I want to be an advocate for its adoption at work, but it's so difficult to understand what's happening or why some strategies don't work very well, that we're very close to abandoning its use. I'm hoping someone can help point me in the right direction!

I had a similar experience at a previous company. In theory, this library should be the greatest thing ever. But in reality, we found that for any slightly nontrivial use case, `hypothesis` simply never seemed to work out of the box. Tracking down the causes of flaky stateful tests and tuning timeout deadlines was exceedingly painful. The false positive rate just goes through the roof, since the vast majority of test failures in my experience were basically just "`hypothesis` doesn't like how you wrote/configured this" rather than "`hypothesis` found a bug".

**Comment by honno** [comment]
Yeah not useful if the repr isn't saying what assumption isn't being satisfied. Not much familiar with the health check system, hopefully someone else can shine a light if things can be improved.

Just a quick note—do you need `fill=st.nothing()`? Examples might not be as interesting with the default `fill=None`, but they're still plenty interesting in my experience, and definitely scales better for such large dfs.

Although generally, as you've found, Hypothesis doesn't do great with such dimensional data. Personally I'd typically use Hypothesis just for generating relatively-small (but interesting) examples when integration testing, where I'm more likely to lean on hard-coded,  maybe-[parametrized](https://docs.pytest.org/en/6.2.x/parametrize.html) tests otherwise for say large and relational inputs. Satisfying health checks (i.e. not suppressing them via `@settings`) is usually my rubric for how I purposely restrict my use of Hypothesis strategies.

**Comment by langfield** [comment]
> Satisfying health checks (i.e. not suppressing them via @settings) is usually my rubric for how I purposely restrict my use of Hypothesis strategies.

Neat! I like this heuristic!

**Comment by epgui** [comment]
> Just a quick note—do you need fill=st.nothing()? Examples might not be as interesting with the default fill=None, but they're still plenty interesting in my experience, and definitely scales better for such large dfs.

I had the same issue before I started passing the default fill strategy. The only reason I added this is because the docs say:

> fill is a strategy that may be used to generate a single background value for the array. If None, a suitable default will be inferred based on the other arguments. If set to nothing() then filling behaviour will be disabled entirely and every element will be generated independently.

And it sounded to me like maybe passing in nothing() would improve performance / require less checks. I did not measure if this made a difference in performance, but it didn't seem to help with the error I'm getting.

---

> Tracking down the causes of flaky stateful tests and tuning timeout deadlines was exceedingly painful.

Just for extra context, our code is written in a FP style with mostly pure functions and no shared state (which I highly recommend as a general practice, with few exceptions). Hypothesis adds a huge layer of statefulness that stretches my own reasoning / debugging abilities (tone check: I hope this comes across as humility, because I really appreciate this library... I just wish I could use it more!).

**Comment by langfield** [comment]
> Just for extra context, our code is written in a FP style with mostly pure functions and no shared state (which I highly recommend as a general practice, with few exceptions).

That's interesting! I've tried to do this in python on a few occasions and had difficulty. What do you use to do IO and error handling?

**Comment by epgui** [comment]
> What do you use to do IO and error handling?

I don't want to go too off-topic, but nothing special really, we don't do pure FP in python: I don't necessarily recommend "writing haskell in python". We throw and catch regular exceptions where appropriate, and we do normal IO. But we make a conscious effort to not do any mutation when we don't absolutely need to, to push side effects to the boundaries of our application, to not use classes unless we really need to, etc. We write small, simple, functional-style wrappers around libraries like boto or pygsheets. We make use of functools.reduce (I'm sorry Guido, but it was a bad idea to move that out of the way) and a bunch of utility functions from [pytoolz](https://toolz.readthedocs.io/en/latest/), encourage the use of function composition, currying, list comprehensions... Discourage the use of for loops generally (exceptions apply). We don't try to be perfect, but we try to make an effort.

**Comment by langfield** [comment]
> I don't want to go too off-topic

Ah, yes, apologies for that. Entirely my bad! 😬  I hadn't heard of `pytoolz` or `fn.py`, thanks for the link! The currying and Scala-style placeholder lambdas are particularly cool. I have a ton of code that looks like this:
```python
    # Write cards and models to disk for each deck.
    write = partial(write_deck, col, targetdir, colnotes, tidy_field_files)
    notefiles: Dict[NoteId, CardFileMap] = reduce(write, decks, {})
    _: Set[Optional[T]] = set(map(partial(write_models, col, models), decks))

    # Get all POSIX-style symlinks created on Windows.
    cardfiles = F.cat(F.cat(map(lambda x: x.values(), notefiles.values())))
    links: Iterable[Optional[WindowsLink]] = map(lambda c: c.link, cardfiles)
    windows_links: Set[WindowsLink] = set(filter(lambda l: l is not None, links))
```
And those features would make it a lot more concise!

**Comment by rsokl** [comment]
Apologies on the slow feedback on this. Your minimal example is great. Thank you for taking the time to write this up ❤️ 

> it sounded to me like maybe passing in nothing() would improve performance / require less checks

The fill strategy is to enable Hypothesis to draw a single value for the majority of array elements. Disabling it means that, as the docs say, "every element will be generated independently". Thus each value has to be individually tracked/shrunk. This will deteriorate strategy health and runtime performance.


> I've been trying to adopt hypothesis to generate plausible data sets

Hypothesis is primarily meant to test properties of your code. These can be things like invariance to data-ordering, not crashing on empty datasets, not crashing on nans/infs, idempotence, etc. It is typically the case that you do not need large-sized datasets to test such properties.

If you want to pass a large-sized examples into integration tests, then you might consider using pytest fixtures to accomplish this. You can even use your strategy to manually generate a handful of large-ish examples, and then copy them (statically) into a fixture.

> I've reproduced below a typical, minimal example of what I'm trying to do: 

With regard to your specific example, moving your filter to the row that you care about alleviates the `too_slow` healtcheck (at least on my machine), and letting hypothesis use a fill strategy lets me remove the other two health checks. That being said, I do need to keep `max_examples` to ~5.

```python


@st.composite
def my_data_set_strategy(draw, num_rows):
    cols = [
        # Not even bothering with strings yet, I figure integers would be faster
        column('company_name', elements=st_integers, dtype=int),
        column('date', elements=gen_dt_year_month()),
        column('field_a', elements=st_integers.filter(bool), fill=st.nothing(), dtype=int),
        column('field_b', elements=st_integers, dtype=int),
        column('field_c', elements=st_integers, dtype=int),
        column('field_d', elements=st_integers, dtype=int),
        column('field_e', elements=st_integers, dtype=int),
        column('field_f', elements=st_integers, dtype=int),
        column('field_g', elements=st_integers, dtype=int),
        column('field_h', elements=st_integers, dtype=int),
        column('field_i', elements=st_integers, dtype=int),
        column('field_j', elements=st_integers, dtype=int),
        column('field_k', elements=st_integers, dtype=int),
        column('field_l', elements=st_integers, dtype=int),
        column('field_m', elements=st_integers, dtype=int),
        column('field_n', elements=st_integers, dtype=int),
        column('field_o', elements=st_integers, dtype=int),
        column('field_p', elements=st_integers, dtype=int),
        column('field_q', elements=st_integers, dtype=int),
        column('field_r', elements=st_integers, dtype=int),
        column('field_s', elements=st_integers, dtype=int),
        column('field_t', elements=st_integers, dtype=int),
        column('field_u', elements=st_integers, dtype=int),
        column('field_v', elements=st_integers, dtype=int),
        column('field_w', elements=st_integers, dtype=int),
        column('field_x', elements=st_integers, dtype=int),
        column('field_y', elements=st_integers, dtype=int),
        column('field_z', elements=st_integers, dtype=int),
    ]
    return draw(
        data_frames(
            columns=cols,
            index=range_indexes(min_size=num_rows, max_size=num_rows),
        ).filter(
            # Eliminate dataframes that have all-zero values because those are not
            # important to consider in the context of our tests.
            lambda df: set(df["field_a"]) != {0}
        )
    )


@given(my_data_set_strategy(num_rows=50))
@settings(
    # Turning this down aggressively for performance reasons
    max_examples=5,
    # Please tell me more!
    verbosity=Verbosity.verbose,
)
def thing(data):
    print(data)
    assert True
```

> It's unclear to me what assumption it's unable to satisfy: the only constraint I have is for all values in the dataframe to not be zero-valued, which I would expect would be no problem after the first handful (<5) of examples.

In your version of the strategy, you are asking Hypothesis to draw the full dataframe, inspect it for zeroes in col-A, and throw it all away if you see one. Given that the `integers` strategy shrinks to 0, this is a common occurrence. You are basically asking Hypothesis draw 1300 values and deduce what *combination* of values may have lead it to produce a dataframe that failed assumption – Hypothesis has no way to know that you are just looking at individual elements in column A – hence the `Unsatisfiable` error. If you move the filter to the specific column, Hypothesis will know that it needs to filter zero for any element in that column – this is *much* more feasible.

> I really want to love this library, and I want to be an advocate for its adoption at work, but it's so difficult to understand what's happening or why some strategies don't work very well, that we're very close to abandoning its use.

We want you to love it too! I would love to figure out what we can distill from this issue – perhaps improved docs, a blog post, a youtube video?? – that could help people in more of a data science or ML role how to use Hypothesis for their purposes. I am a ML researcher and engineer and use Hypothesis for my projects all of the time. It certainly does take time to develop and instinct for good/bad practices with Hypothesis, and I would love to shorten that amount of time for people.

I am closing this issue, but please feel free to follow up here with any questions or thoughts that you have, and sorry for having to deal with this frustrating issue.

**Comment by epgui** [comment]
@rsokl First, thank you very much for your kind and detailed response! There is one thing that I really don't understand, however...

> you are asking Hypothesis to draw the full dataframe, inspect it for zeroes in col-A, and throw it all away if you see one

In `.filter(lambda df: set(df["field_a"]) != {0})`, is it not correct to say that we're filtering out dataframes that contain **only** zeros and nothing else in col-A? We're checking if the **set** of all values is == {0}. If there was **at least one** value that was non-zero, then wouldn't the condition be satisfied?

In other words, it's totally okay if there are **some** zeros in column A: I just don't want **all** the values to be zero.

**Comment by rsokl** [comment]
> is it not correct to say that we're filtering out dataframes that contain only zeros and nothing else in col-A?

Ah, yes. That is correct! I misread the filter expression. My apologies.

In that case, the solution I provided is *not* equivalent to what you posted. The issue with your filter being a challenge for Hypothesis to fulfill still stands for the reasons I provided; perhaps you could provide `fill=st.sampled_from(range(1, 1000, 10))` as a magnitude-hopping, "cheat" to avoid all-zeros.

**Comment by epgui** [comment]
If what you say is true, then my mental model for thinking about this is wrong: my intuition is that the probability of drawing all-zeros many many many times in a row would be close to p=0.0, even if the probability of drawing all-zeros on the first draw was close to p=1.0.

But now that you point this out, I wonder... Does the shrinking happen "bottom-up" / "depth-first", or "top-down" / "breadth-first"?

In other words... If I set a condition on the dataframe, and the generated dataframe fails to meet this condition, then I would not expect the integer strategy to do any shrinking right away: rather, I would expect the shrinking to happen first at the level of the dataframe or column, and then at the level of its children (this is what I would call "top-down" or "breadth-first" shrinking, and is how I would assume this would work).

Am I correct in thinking that this is where my understanding is wrong?

**Comment by Zac-HD** [comment]
> @rsokl If what you say is true, then my mental model for thinking about this is wrong: my intuition is that the probability of drawing all-zeros many many many times in a row would be close to p=0.0, even if the probability of drawing all-zeros on the first draw was close to p=1.0.

Hypothesis' internals induce all kinds of weird correlations within (and between) examples precisely *because* this sort of edge case is disproportionally likely to trigger bugs.  It's unfortunate that this makes reasoning about the distribution of examples so difficult.

> But now that you point this out, I wonder... Does the shrinking happen "bottom-up" / "depth-first", or "top-down" / "breadth-first"? ... Am I correct in thinking that this is where my understanding is wrong?

It's rather complicated: I'd suggest reading David's paper https://drops.dagstuhl.de/opus/volltexte/2020/13170/ and https://github.com/HypothesisWorks/hypothesis/tree/master/guides on internals and shrinking if you'd like to know more.  "Smart edits to a tree structure" is probably the best short approximation, though, and usually that means trying to remove large chunks as early as possible.

**Comment by epgui** [comment]
@Zac-HD I'm not sure where that leaves us for using hypothesis... But you have both certainly been very helpful.

I'll take a look at these, and if I'm brave enough, the source code, and compare it to clojure/spec/testcheck's. Something feels off about hypothesis' behaviour here, but I don't know enough to say for sure.

**Comment by Zac-HD** [comment]
If you're going to look into Hypothesis' internals, I recommend starting with https://github.com/DRMacIver/minithesis to get a good idea of the basic principles underlying the engine as well as the links above.
